kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,771 @@
+local cjson = require("cjson.safe")
+local utils = require("kong.resty.dns_client.utils")
+local mlcache = require("kong.resty.mlcache")
+local resolver = require("resty.dns.resolver")
+
+local now       = ngx.now
+local log       = ngx.log
+local ERR       = ngx.ERR
+local WARN      = ngx.WARN
+local DEBUG     = ngx.DEBUG
+local ALERT     = ngx.ALERT
+local timer_at  = ngx.timer.at
+
+local type            = type
+local pairs           = pairs
+local ipairs          = ipairs
+local math_min        = math.min
+local string_lower    = string.lower
+local table_insert    = table.insert
+local table_isempty   = require("table.isempty")
+local tablex_readonly = require("pl.tablex").readonly
+
+local parse_hosts   = utils.parse_hosts
+local ipv6_bracket  = utils.ipv6_bracket
+local search_names  = utils.search_names
+local get_next_round_robin_answers     = utils.get_next_round_robin_answers
+local get_weighted_round_robin_answers = utils.get_weighted_round_robin_answers
+
+local req_dyn_hook_run_hooks = require("kong.dynamic_hook").run_hooks
+
+
+-- Constants and default values
+
+local DEFAULT_ERROR_TTL   = 1     -- unit: second
+local DEFAULT_STALE_TTL   = 4
+local DEFAULT_EMPTY_TTL   = 30
+-- long-lasting TTL of 10 years for hosts or static IP addresses in cache settings
+local LONG_LASTING_TTL    = 10 * 365 * 24 * 60 * 60
+
+local PERSISTENT_CACHE_TTL = { ttl = 0 }  -- used for mlcache:set
+
+local DEFAULT_ORDER = { "LAST", "SRV", "A", "AAAA", "CNAME" }
+
+local TYPE_SRV      = resolver.TYPE_SRV
+local TYPE_A        = resolver.TYPE_A
+local TYPE_AAAA     = resolver.TYPE_AAAA
+local TYPE_CNAME    = resolver.TYPE_CNAME
+local TYPE_LAST     = -1
+
+local NAME_TO_TYPE = {
+  SRV     = TYPE_SRV,
+  A       = TYPE_A,
+  AAAA    = TYPE_AAAA,
+  CNAME   = TYPE_CNAME,
+  LAST    = TYPE_LAST,
+}
+
+local TYPE_TO_NAME = {
+  [TYPE_SRV]      = "SRV",
+  [TYPE_A]        = "A",
+  [TYPE_AAAA]     = "AAAA",
+  [TYPE_CNAME]    = "CNAME",
+  [TYPE_LAST]     = "LAST",
+}
+
+local HIT_L3 = 3 -- L1 lru, L2 shm, L3 callback, L4 stale
+
+local HIT_LEVEL_TO_NAME = {
+  [1] = "hit_lru",
+  [2] = "hit_shm",
+  [3] = "hit_cb",
+  [4] = "hit_stale",
+}
+
+-- server replied error from the DNS protocol
+local NAME_ERROR_CODE             = 3 -- response code 3 as "Name Error" or "NXDOMAIN"
+-- client specific error
+local CACHE_ONLY_ERROR_CODE       = 100
+local CACHE_ONLY_ERROR_MESSAGE    = "cache only lookup failed"
+local CACHE_ONLY_ANSWERS = tablex_readonly({ errcode = CACHE_ONLY_ERROR_CODE, errstr = CACHE_ONLY_ERROR_MESSAGE })
+local EMPTY_RECORD_ERROR_CODE     = 101
+local EMPTY_RECORD_ERROR_MESSAGE  = "empty record received"
+
+
+-- APIs
+
+local _M = {}
+local mt = { __index = _M }
+
+-- copy TYPE_*
+for k,v in pairs(resolver) do
+  if type(k) == "string" and k:sub(1,5) == "TYPE_" then
+    _M[k] = v
+  end
+end
+_M.TYPE_LAST = TYPE_LAST
```

Comment:
```suggestion
local _M = {
  TYPE_SRV     = TYPE_SRV,
  TYPE_A       = TYPE_A,
  TYPE_AAAA    = TYPE_AAAA,
  TYPE_CNAME   = TYPE_CNAME,
  TYPE_LAST    = TYPE_LAST,
}
local mt = { __index = _M }
```

The dynamic copy was only intended to reduce maintenance, but since we already list all the types several times for all the lookup/conversion tables, we can list them here again. This improves readability over the dynamic copy code.